google/cloud/spanner/samples/samples.cc

```diff
@@ -1160,6 +1125,208 @@ void DeleteData(google::cloud::spanner::Client client) {
+    std::cout << "PopularityScore: " << std::get<2>(*row).value() << "\n";
+  }
+  std::cout << "Query completed for [spanner_query_with_float_parameter]\n";
+}
+//! [END spanner_query_with_float_parameter]
+
+//! [START spanner_query_with_int_parameter]
+void QueryWithIntParameter(google::cloud::spanner::Client client) {
+  namespace spanner = ::google::cloud::spanner;
+  int example_int = 3000;
+  spanner::SqlStatement select(
+      "SELECT VenueId, VenueName, Capacity FROM Venues"
+      " WHERE Capacity >= @capacity",
+      {{"capacity", spanner::Value(example_int)}});
+  using RowType = std::tuple<std::int64_t, absl::optional<std::string>,
+                             absl::optional<std::int64_t>>;
+  auto rows = client.ExecuteQuery(std::move(select));
+  for (auto const& row : spanner::StreamOf<RowType>(rows)) {
+    if (!row) throw std::runtime_error(row.status().message());
+    std::cout << "VenueId: " << std::get<0>(*row) << "\t";
+    std::cout << "VenueName: " << std::get<1>(*row).value() << "\t";
+    std::cout << "Capacity: " << std::get<2>(*row).value() << "\n";
+  }
+  std::cout << "Query completed for [spanner_query_with_int_parameter]\n";
+}
+//! [END spanner_query_with_int_parameter]
+
+//! [START spanner_query_with_string_parameter]
+void QueryWithStringParameter(google::cloud::spanner::Client client) {
+  namespace spanner = ::google::cloud::spanner;
+  std::string example_string = "Venue 42";
+  spanner::SqlStatement select(
+      "SELECT VenueId, VenueName FROM Venues"
+      " WHERE VenueName = @venue_name",
+      {{"venue_name", spanner::Value(example_string)}});
+  using RowType = std::tuple<std::int64_t, absl::optional<std::string>>;
+  auto rows = client.ExecuteQuery(std::move(select));
+  for (auto const& row : spanner::StreamOf<RowType>(rows)) {
+    if (!row) throw std::runtime_error(row.status().message());
+    std::cout << "VenueId: " << std::get<0>(*row) << "\t";
+    std::cout << "VenueName: " << std::get<1>(*row).value() << "\t";
+  }
+  std::cout << "Query completed for [spanner_query_with_string_parameter]\n";
+}
+//! [END spanner_query_with_string_parameter]
+
+//! [START spanner_query_with_timestamp_parameter]
+void QueryWithTimestampParameter(google::cloud::spanner::Client client) {
+  namespace spanner = ::google::cloud::spanner;
+  auto example_timestamp =
+      spanner::MakeTimestamp(std::chrono::system_clock::now()).value();
+  spanner::SqlStatement select(
+      "SELECT VenueId, VenueName, LastUpdateTime FROM Venues"
+      " WHERE LastUpdateTime < @last_update_time",
+      {{"last_update_time", spanner::Value(example_timestamp)}});
+  using RowType = std::tuple<std::int64_t, absl::optional<std::string>,
+                             absl::optional<spanner::Timestamp>>;
+  auto rows = client.ExecuteQuery(std::move(select));
+  for (auto const& row : spanner::StreamOf<RowType>(rows)) {
+    if (!row) throw std::runtime_error(row.status().message());
```

Comment:
This code is also not reached.  Which is odd given that we are using "now" the data was inserted in the past.  Maybe we should consider `now() + seconds(20)` or something?

Reply:
I see it finding matches (see below), but I agree that it could well be a race.

The Go sample changes the inequality to `<=`.  I'll do the same.

```
Running spanner_query_with_timestamp_parameter sample
2020-09-03T18:09:18.906254766Z [DEBUG] ExecuteStreamingSql() << session: "projects/cloud-cpp-testing-resources/instances/test-instance-11/databases/db-2020-09-03-x3r0pc-u4u8y_6pu/sessions/AN4G3x8CElBR9qMSfrkwTLINfvv8R7QnfmvFDxrUZiaA0-aAyai1xZ_UIZ6LuQ"
transaction {
  single_use {
    read_only {
      strong: true
      return_read_timestamp: true
    }
  }
}
sql: "SELECT VenueId, VenueName, LastUpdateTime FROM Venues WHERE LastUpdateTime < @last_update_time"
params {
  fields {
    key: "last_update_time"
    value {
      string_value: "2020-09-03T18:09:18.905830539Z"
    }
  }
}
param_types {
  key: "last_update_time"
  value {
    code: TIMESTAMP
  }
}
seqno: 1
 (./google/cloud/internal/log_wrapper.h:102)
2020-09-03T18:09:18.906686775Z [DEBUG] ExecuteStreamingSql() >> not null stream (./google/cloud/internal/log_wrapper.h:104)
VenueId: 4      VenueName: Venue 4      LastUpdateTime: 2020-09-03T18:09:18.299631Z
VenueId: 19     VenueName: Venue 19     LastUpdateTime: 2020-09-03T18:09:18.299631Z
VenueId: 42     VenueName: Venue 42     LastUpdateTime: 2020-09-03T18:09:18.299631Z
Query completed for [spanner_query_with_timestamp_parameter]
```